internal/driver/common/canvas.go

```diff
@@ -288,6 +304,25 @@ func (c *Canvas) Painter() gl.Painter {
 
 // Refresh refreshes a canvas object.
 func (c *Canvas) Refresh(obj fyne.CanvasObject) {
+	walkNeeded := false
+	switch obj.(type) {
+	case *fyne.Container:
+		walkNeeded = true
+	case fyne.Widget:
+		walkNeeded = true
+	}
+
+	if walkNeeded {
+		c.serializeRefresh.Lock()
+		driver.WalkCompleteObjectTree(obj, func(co fyne.CanvasObject, p1, p2 fyne.Position, s fyne.Size) bool {
+			if i, ok := co.(*canvas.Image); ok {
+				i.Refresh()
+			}
+			return false
+		}, nil)
+		c.serializeRefresh.Unlock()
```

Comment:
If we are serializing the refresh call, do we need a benchmark to indicate performance changes here?

Reply:
What would you measure in the benchmark? Past behavior was loading image in the render thread which would have generated frame drop there, but this is now moved to the user thread. So less framedrop, but impact on user thread.

Also we do not have the ability to know when a change hit the screen, so measuring the time between a Refresh to when we get the change to screen is difficult. Do you have any suggestion maybe?

Reply:
I am a bit disconnected from the problem we are trying to solve here. Is there a code snippet to reproduce the "framedrop" issue? It sounds to me that this is a decision tradeoff regarding higher performance but not fully correct or lower performance but fully correct. I think, if the "framedrop" only happens during loading, then the optimization can be done a bit better and guarantee the "lock" behavior only happens during loading. To benchmark, it would consider two cases: 1) refresh without loading; 2) refresh during loading.

Reply:
For a "framedrop" to happen, it would have to be in the render thread, but the idea is to move all image.Refresh call outside of the render thread. As the image.Refresh is now outside of the render thread, there is no loading during the rendering of the frame.

So with the image.Refresh always happening outside of the render thread, it means the loading is not happening in the same place as with past version of Fyne. This logically move the time spend in loading image from the render thread to the thread calling Refresh. As the loading is done before notifying the render thread, there is no framedrop as the image loading is not blocking the rendering thread. Still the total time spend on the task remain the same as long as there is no excess call to Refresh.

With that in mind, I do not see what to benchmark as we do not have the ability to know when a change outside of the render thread has been rendered. If we measure only the Refresh time, it won't be comparable as we moved where the computation is happening.

Reply:
Discussed on the call this change does make sense and it is hard to show in benchmarking.
We agreed that it is part of a wider improvement to how `Canvas.Refresh` is used as well so we can lead to better performance overall.

Reply:
I looked up the notes [here](https://github.com/fyne-io/meetings/blob/main/2023/2023-04-14.md#image-render-on-event-thread-3785) and agreed that it would be helpful to document the practices for calling `Refresh`